tests/mem_safety/segment-free.wast

```diff
@@ -0,0 +1,21 @@
+(module
+  (memory 1 1)
+  (global $sp (mut i32) (i32.const 64))
+  (func (export "foo") (result i32)
+        (local $ptr i64)
+        (i32.const 0)
+        (i32.const 16)
+        (segment.stack_new)
+        (local.tee $ptr)
+        (i32.const 42)
+        (i32.store_segment)
+
+        (local.get $ptr)
+        (i32.load_segment)
+
+        (local.get $ptr)
+        (i32.const 16)
+        (segment.free)
+        (return)
+  )
+)
```

Comment:
You could change the test case to the following:

```suggestion
(module
  (memory 1 1)
  (global $sp (mut i32) (i32.const 64))
  (func (export "foo") (param $idx i64) (result i32)
    (local $ptr i64)
    (i32.const 0)
    (i32.const 16)
    (segment.stack_new)
    (local.tee $ptr)
    (local.get $idx)
    (i64.const 2)
    (i64.shl)
    (i64.add)
    (i32.const 42)
    (i32.store_segment)

    (local.get $ptr)
    (local.get $idx)
    (i64.const 2)
    (i64.shl)
    (i64.add)
    (i32.load_segment)

    (local.get $ptr)
    (i32.const 16)
    (segment.free)
    (return)
  )
)
```

This basically just adds a parameter, so we can access the memory with a user-supplied index and check if memory protection works in the following way:

```shell
$ ./wasmtime run --allow-precompiled --cranelift-enable use_mte --wasm-features=mem-safety segment-free.cwasm --invoke foo 3
42
$ ./wasmtime run --allow-precompiled --cranelift-enable use_mte --wasm-features=mem-safety segment-free.cwasm --invoke foo 4
Caused by:
    0: failed to invoke `foo`
    1: error while executing at wasm backtrace:
           0:   0x45 - <unknown>!<wasm function 0>
    2: memory fault at wasm address 0x10 in linear memory of size 0x10000
    3: wasm trap: out of bounds memory access
```